wellcomecollection/aws-account-infrastructure
# use oidc to give repo permissions needed to run formatting

APPLIED 

## What does this change?

- Adds a role to pull images from the platform account ecr
- Allows catalogue-api repo to assume the role so that a github workflow can pull images from the ecr

## How to test

Once applied, push a change to the `test-autoformat-github-workflow` branch of the catalogue-api repo
See if the workflow can successfully run formatting

## How can we measure success?

The `autoformat.yml` workflow can successfully run in the catalogue-api 
At this point, it should not run successfully in any other repos to which the permission doesn't extend

## Have we considered potential risks?

The `github_repo_role` tf limits permissions to the repo listed in `accounts/platform/github_oidc.tf` so should be safe.
Can easily be destroyed if necessary

`terraform plan`
```
Terraform will perform the following actions:

  # data.aws_iam_policy_document.platform_ecr_read_assume_role_policy will be read during apply
  # (config refers to values not yet known)
 <= data "aws_iam_policy_document" "platform_ecr_read_assume_role_policy" {
      + id   = (known after apply)
      + json = (known after apply)

      + statement {
          + actions   = [
              + "sts:AssumeRole",
            ]
          + resources = [
              + (known after apply),
            ]
        }
    }

  # aws_iam_role.platform_ecr_read will be created
  + resource "aws_iam_role" "platform_ecr_read" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRole"
                      + Effect    = "Allow"
                      + Principal = {
                          + AWS = "arn:aws:iam::760097843905:root"
                        }
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = false
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 3600
      + name                  = (known after apply)
      + name_prefix           = (known after apply)
      + path                  = "/"
      + tags_all              = {
          + "TerraformConfigurationURL" = "https://github.com/wellcomecollection/aws-account-infrastructure/tree/main/accounts/platform"
        }
      + unique_id             = (known after apply)

      + inline_policy (known after apply)
    }

  # aws_iam_role_policy.platform_ecr_read will be created
  + resource "aws_iam_role_policy" "platform_ecr_read" {
      + id          = (known after apply)
      + name        = (known after apply)
      + name_prefix = (known after apply)
      + policy      = jsonencode(
            {
              + Statement = [
                  + {
                      + Action   = [
                          + "ecr:GetDownloadUrlForLayer",
                          + "ecr:BatchGetImage",
                        ]
                      + Effect   = "Allow"
                      + Resource = "760097843905.dkr.ecr.eu-west-1.amazonaws.com/*"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + role        = (known after apply)
    }

  # module.platform_ecr_read_assume_role.aws_iam_role.github_actions will be created
  + resource "aws_iam_role" "github_actions" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRoleWithWebIdentity"
                      + Condition = {
                          + StringEquals = {
                              + "token.actions.githubusercontent.com:aud" = "sts.amazonaws.com"
                            }
                          + StringLike   = {
                              + "token.actions.githubusercontent.com:sub" = "repo:wellcomecollection/catalogue-api:*"
                            }
                        }
                      + Effect    = "Allow"
                      + Principal = {
                          + Federated = "arn:aws:iam::760097843905:oidc-provider/token.actions.githubusercontent.com"
                        }
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + description           = "Inital role assumed by the GitHub OIDC provider"
      + force_detach_policies = false
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 3600
      + name                  = (known after apply)
      + name_prefix           = "gha-platform_ecr_read-"
      + path                  = "/"
      + tags_all              = {
          + "TerraformConfigurationURL" = "https://github.com/wellcomecollection/aws-account-infrastructure/tree/main/accounts/platform"
        }
      + unique_id             = (known after apply)

      + inline_policy (known after apply)
    }

  # module.platform_ecr_read_assume_role.aws_iam_role_policy.github_actions_assume_role_policy will be created
  + resource "aws_iam_role_policy" "github_actions_assume_role_policy" {
      + id          = (known after apply)
      + name        = (known after apply)
      + name_prefix = (known after apply)
      + policy      = (known after apply)
      + role        = (known after apply)
    }
```
